### PR TITLE
Add NetworkInterface.c for TMS320F2838x

### PR DESCRIPTION
**Pull request by adrien-cardinale**
# NetworkInterface.c for TMS320F2838x

Description
-----------
Add NetworkInterface.c for TMS320F2838x

Test Steps
-----------
Test with TMS320F28388d from texas instrument

Checklist:
----------

- [ ] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by htibosch** [comment]
@adrien-cardinale , thank you for this PR.

Could you write the text "/bot run formatting" (without the quotes) and send it as a comment. The source code will be formatted to the FreeRTOS standard. After that you can do a pull on your own PR to get the same formatting on your local repo.

File: source/portable/NetworkInterface/TMS320F2838x/NetworkInterface.c [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+
+    uint32_t ret = Ethernet_sendPacket(emac_handle, &pktDesc);
+
+    if(ret == 0){
+        vReleaseNetworkBufferAndDescriptor(pxDescriptor);
+    }
+
+    if(xReleaseAfterSend != pdFALSE){
+        pxDescriptor->pucEthernetBuffer = NULL;
+        vReleaseNetworkBufferAndDescriptor(pxDescriptor);
```
Comment (htibosch) [inline source/portable/NetworkInterface/TMS320F2838x/NetworkInterface.c:257]: In the case when `if(ret == 0)`, you have already called `vReleaseNetworkBufferAndDescriptor()`. It would cause an erro to call it again.

File: source/portable/NetworkInterface/TMS320F2838x/NetworkInterface.c [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+    IPStackEvent_t xRxEvent;
+    uint32_t ulISREvents = 0U;
+    for( ;; )
+    {
+        xTaskNotifyWait( 0U,
+                         EMAC_IF_ALL_EVENT,
+                         &( ulISREvents ),
+                         portMAX_DELAY );
+        if((ulISREvents & EMAC_IF_RX_EVENT) != 0){
+            pxBufferDescriptor = pxGetNetworkBufferWithDescriptor( ETHERNET_MAX_PACKET_LENGTH, 0 );
```
Comment (htibosch) [inline source/portable/NetworkInterface/TMS320F2838x/NetworkInterface.c:273]: Here you are handling a `EMAC_IF_RX_EVENT`, meaning that a packet has been received. Wouldn't you check the validity of `pPktDescGlobal` here, see if `validLength` is within range and so?
Not need to copy `ETHERNET_MAX_PACKET_LENGTH` bytes, that would be a waste of time.
Also I wouldn't create a packet buffer before calling `eConsiderFrameForProcessing()` and `FreeRTOS_MatchingEndpoint()`: both functions might fail.
Have a look at this:
~~~
if( valid_packet( pPktDescGlobal )
{
    /* Check if we should handle this packet. */
    if( eConsiderFrameForProcessing( pPktDescGlobal->dataBuffer ) == eProcessBuffer )
    {
        /* Find the proper endpoint. */
        NetworkEndPoint_t * pxEndPoint = FreeRTOS_MatchingEndpoint( pxMyInterface, pPktDescGlobal->dataBuffer );

        if( pxEndPoint != NULL )
        {
            NetworkBufferDescriptor_t * pxBufferDescriptor = pxGetNetworkBufferWithDescriptor( pPktDescGlobal->validLength, 0 );
            memcpy( pxBufferDescriptor->pucEthernetBuffer, pPktDescGlobal->dataBuffer, pPktDescGlobal->validLength );
            pxBufferDescriptor->xDataLength = pPktDescGlobal->validLength;

            pxBufferDescriptor->pxEndPoint = pxEndPoint;
            xRxEvent.eEventType = eNetworkRxEvent;
            xRxEvent.pvData = ( void * ) pxBufferDescriptor;

            if( xSendEventStructToIPTask( &xRxEvent, 0 ) == pdTRUE )
            {
                iptraceNETWORK_INTERFACE_RECEIVE();
            }
            else
            {
                iptraceETHERNET_RX_EVENT_LOST();
                vReleaseNetworkBufferAndDescriptor( pxBufferDescriptor );
            }
        }
        else
        {
            iptraceETHERNET_RX_EVENT_LOST();
        }
    }
    else
    {
        iptraceETHERNET_RX_EVENT_LOST();
    }
}
~~~
In other words, do not call `pxGetNetworkBufferWithDescriptor()` until you are sure that it will be used.

Question: How do we make sure that `pPktDescGlobal` isn't accessed before by interrupt code before we've handled the last packet?
I am thinking of adding a **queue** or a **counting semaphore** between the EMAC task and the callback functions:
~~~
Ethernet_receivePacketCallbackCustom()
Ethernet_releaseTxPacketBufferCallbackCustom()
~~~
A `EMAC_IF_RX_EVENT` event means: one or more packets have been received.
A `EMAC_IF_TX_EVENT` event means the one or more packets have been sent.
But while you are handling some `EMAC_IF` event, the interrupt may get triggered.

File: source/portable/NetworkInterface/TMS320F2838x/NetworkInterface.c [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+        xTaskCreate(prvEMACDeferredInterruptHandlerTask, "EMACInt", configMINIMAL_STACK_SIZE, NULL, configMAX_PRIORITIES - 1, &xEMACTaskHandle);
+        configASSERT(xEMACTaskHandle);
+        pxMyInterface = pxInterface;
+    }
+
+    return xReturn;
+}
+
+static BaseType_t prvf28388d_NetworkInterfaceOutput(NetworkInterface_t* pxInterface, NetworkBufferDescriptor_t* const pxDescriptor, BaseType_t xReleaseAfterSend){
+    pxGlobalDescriptor = pxDescriptor;
```
Comment (htibosch) [inline source/portable/NetworkInterface/TMS320F2838x/NetworkInterface.c:235]: You are saving the packet buffer because `prvEMACDeferredInterruptHandlerTask()` will release it later on.
Please check `xReleaseAfterSend` : when false, you may not release the buffer.
And also, a few lines later in this function, you are releasing `pxDescriptor`, without invalidating the copy in `pxGlobalDescriptor`.

**Review by htibosch — COMMENTED** [review]
Hello @adrien-cardinale , I see that you already made changes after my more [informal review](https://forums.freertos.org/t/networkinterface-for-tms320f2838x-cm/18585/5) on the forum.
In this review a few more proposals that should help making it more efficient.
Also I am proposing to think about the 2 streams: TX packets that have been sent, and RX packets that have been received.

**Comment by AniruddhaKanhere** [comment]
@adrien-cardinale to fix the spell check, please add the words highlighted in the spell check [report](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/actions/runs/7200277559/job/19613856029?pr=1052#step:3:1454) to the [.cSpellWords.txt](https://github.com/adrien-cardinale/FreeRTOS-Plus-TCP/blob/main/.github/.cSpellWords.txt).

**Comment by htibosch** [comment]
Hello @adrien-cardinale, how is the driver going? I wonder if there is anything we can help you with? Is there anything unclear?
Or are you too busy to get on with it?

**Comment by adrien-cardinale** [comment]
Hello @htibosch , thank you for your interest and comments. I'm actually too busy at the moment to make the changes you suggested. This is why I wanted to publish the code as is so that it can serve as a basis for other people who would like to use the driver for the TMS320F2838x.

**Comment by amazonKamath** [comment]
@adrien-cardinale - Thanks for your efforts on this PR. To prevent stale PR's, we'll be closing this for now. Feel free to reopen when you're ready to address the comments.
